4.x: Concurrency limits module, and support in Helidon WebServer

tomas-langer opened the pull request:
### Description
Resolves #8897
Resolves #9229

### Documentation
This PR introduces 
- a new module `common/concurrency/limits` that provides API and SPI for concurrency limit implementations, and a couple of default implementations (AIMD, Fixed).
- a new module `webserver/concurrency-limits` that provides feature with a filter to impose limits within a filter in routing
- update of `webserver/webserver` to use a `Limit` instead of a `Semaphore` in connection handlers (backward compatible)

Configuration reference will be in the generated documentation.

#### Configure limits on WebServer
This will configure limit for a server listener (configurable per listener), enforced on the connection level (i.e. outside of routing and filters).

The default behavior is the same (unlimited).
If `server.max-concurrent-requests` is configured (value is not `-1`), it will be used and `concurrency-limit` configuration for the listener will be ignored.

Configuration:
```yaml
server:
  port: 8080
  concurrency-limit:
    aimd: # `limit type`
      # AIMD limit configuration
```
#### Configure limits for routing
This will configure limit for as a server feature, enforced in an HTTP filter.
Configuration:
```yaml
server:
  features:
    limits: # the feature is called `limits`
      enabled: true
      concurrency-limit: # `limit` configuration of the `limits` server feature
        fixed: # `limit type`
          permits: 1
          queue-length: 10
```

arouel reviewed (approved):
LGTM, great rework and integration

vasanth-bhat commented:
Have few clarifications

1.  Going through    Readme  , there is mention about 'Semaphore' concurrency limit type in  addition to aimd & BulkHead.  Is this same as. "basic" (BasicLimit)?   Also how is this different from specifying.  "server.max-concurrent-requests"? 

2.  For the  BulkHead,   it would be good to have metrics support ,   around  queueing ,  for example  %  queue full.  This is important in production to monitor  the queueing.  

3.  For AMID , it would be good to ahem a metrics around the dynamic concurrency limit in use. A Histogram may be useful here.

tomas-langer commented:
After discussing the needs in Helidon, we must do this using the `permit` approach, so this can be reused in webclient. 
I will refactor the PR.

arouel commented:
I wanted to bring up that metrics would be super important for production as @vasanth-bhat brought it up as well. Within my draft I figured quickly that the limiting works as expected but metrics would give me the insights I need to build convidence. Maybe this is something for a follow-up Pull Request, idk.

tomas-langer commented:
Metrics must be a follow up, as it would require much more work. I cannot just add a dependency on `helidon-metrics-api`, as it would introduce it everywhere (maybe that is the solution, but it requires a bit more thought).
I will finish this PR as the introduction of concurrency limits. We can look into the metrics as a follow up (this would make sense for fault tolerance as well, not just concurrency limits).

tomas-langer commented:
Pushed a new version, updated original description to reflect new implementation.
I have implemented the `fixed` limit as a more general feature - it supports queing now, so we no longer need fault tolerance for this.
How to add this to metrics needs to be a follow up.

tomas-langer commented:
Created a follow up issue: https://github.com/helidon-io/helidon/issues/9304

vasanth-bhat commented:
Went through the new changes.  

  I think it would help to clarify between  the below 2 , with example to understand  the difference between the two.   Do  both them provide same functionality but via different approaches ?

1. Fixed Limit with queueing support at the connection level  (or a server listener )
2. Fixed limit with queueing as server feature enforced via filter.

danielkec commented on common/concurrency/limits/src/main/java/io/helidon/common/concurrency/limits/FixedLimit.java:
```diff
@@ -0,0 +1,348 @@
+
+    private final FixedLimitConfig config;
+    private final LimiterHandler handler;
+    private int initialPermits;
```
Can be final?

arouel commented on webserver/http2/src/main/java/io/helidon/webserver/http2/Http2ServerStream.java:
```diff
@@ -516,15 +516,35 @@ private void handle() {
+                    try {
+                        routing.route(ctx, request, response);
+                    } finally {
+                        if (response.status() == Status.NOT_FOUND_404) {
```
Should we allow a user to specify which 4xx codes can be ignored?

arouel commented on webserver/webserver/src/main/java/io/helidon/webserver/http1/Http1Connection.java:
```diff
@@ -183,30 +187,35 @@ public void handle(Semaphore requestSemaphore) throws InterruptedException {
+                        permit.success();
+                        this.lastRequestTimestamp = DateTime.timestamp();
+                    } catch (Throwable e) {
+                        permit.dropped();
```
Maybe something for a future PR: Should we allow a user to specify which exception can be recorded as ignored instead of being counted as dropped?

tomas-langer commented on common/concurrency/limits/src/main/java/io/helidon/common/concurrency/limits/AimdLimitImpl.java:
```diff
@@ -0,0 +1,193 @@
+        if (!(backoffRatio < 1.0 && backoffRatio >= 0.5)) {
+            throw new ConfigException("Backoff ratio must be within [0.5, 1.0)");
+        }
+        if (maxLimit < minLimit) {
+            throw new ConfigException("Max limit must be higher than min limit");
+        }
+        if (initialLimit > maxLimit) {
+            throw new ConfigException("Initial limit must be lower than max limit");
+        }
+        if (initialLimit < minLimit) {
+            throw new ConfigException("Initial limit must be higher than minimum limit");
+        }
```
ah, the messages are wrong, I see
